enterprise/cmd/frontend/internal/auth/githuboauth/session.go

```diff
@@ -294,3 +294,48 @@ func (s *sessionIssuerHelper) verifyUserOrgs(ctx context.Context, ghClient *gith
 
 	return false
 }
+
+// verifyUserTeams checks whether the authenticated user belongs to one of the GitHub teams listed in the auth.provider > allowOrgsMap configuration
+func (s *sessionIssuerHelper) verifyUserTeams(ctx context.Context, ghClient *githubsvc.V3Client) bool {
+	githubTeams, _, _, err := ghClient.GetAuthenticatedUserTeams(ctx, 1)
+
+	if err != nil {
+		log15.Warn("Could not get GitHub authenticated user teams", "error", err)
+		return false
+	}
+
+	allowedOrgsTeamsFromConfig := make(map[string][]string, len(s.allowOrgsMap))
+	for k, v := range s.allowOrgsMap {
+		allowedOrgsTeamsFromConfig[k] = v
+	}
+
+	for _, ghTeam := range githubTeams {
+		configTeams, ok := allowedOrgsTeamsFromConfig[ghTeam.Organization.Login]
+		if ok {
+			for _, team := range configTeams {
+				if team == ghTeam.Name {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// verifyUserOrgsAndTeams checks if the user belongs to one of the allowed listed orgs or teams provided in the auth.provider configuration.
+func (s *sessionIssuerHelper) verifyUserOrgsAndTeams(ctx context.Context, ghClient *githubsvc.V3Client) (bool, bool) {
+	if len(s.allowOrgs) == 0 && len(s.allowOrgsMap) == 0 {
+		return true, true
+	}
+
+	if len(s.allowOrgs) > 0 && len(s.allowOrgsMap) == 0 {
+		return s.verifyUserOrgs(ctx, ghClient), false
+	}
+
+	if len(s.allowOrgs) == 0 && len(s.allowOrgsMap) > 0 {
+		return false, s.verifyUserTeams(ctx, ghClient)
+	}
+
+	return s.verifyUserOrgs(ctx, ghClient), s.verifyUserTeams(ctx, ghClient)
+}
```

Comment:
Are these two settings intended to be mutually exclusive? I don't think they have to. From my understanding, if a user is _any of_ the org or teams, the check should pass.

```suggestion
// verifyUserOrgsAndTeams checks if the user belongs to one of the allowed listed orgs or teams provided in the auth.provider configuration.
func (s *sessionIssuerHelper) verifyUserOrgsAndTeams(ctx context.Context, ghClient *githubsvc.V3Client) bool {
	if len(s.allowOrgs) == 0 && len(s.allowOrgsMap) == 0 {
		return true
	}

	if len(s.allowOrgs) > 0 && s.verifyUserOrgs(ctx, ghClient) {
		return true
	}

	if len(s.allowOrgsMap) > 0 && s.verifyUserTeams(ctx, ghClient) {
		return true 
	}

	return false
}
```

Reply:
In fact, they are not mutually exclusive. As the function returns 2 values, the check that you are suggesting here is made  above, when the[ function is called](https://github.com/sourcegraph/sourcegraph/pull/35207/files#diff-bfb95b1a576cab8fbb0d4a1edd331dd082ba6e7f563e1d4a300f9197fe7b62a8R64). But I can change it as you suggested to return just one value and move the final check to this function...

Reply:
Hmm, I think there is a problem with the "simpler" version because if there are `s.allowOrgs` set but a user is not part of any, while it is part of `a.allowOrgsMap`, the user won't be able to sign up.

Reply:
You are def right, @unknwon . My mistake. Thanks for catching this 🙏